client/allocrunner/taskrunner/restarts/restarts.go

```diff
@@ -20,9 +20,13 @@ const (
 	ReasonDelay               = "Exceeded allowed attempts, applying a delay"
 )
 
-func NewRestartTracker(policy *structs.RestartPolicy, jobType string) *RestartTracker {
+func NewRestartTracker(policy *structs.RestartPolicy, jobType string, tlc *structs.TaskLifecycleConfig) *RestartTracker {
 	onSuccess := true
-	if jobType == structs.JobTypeBatch {
+	if tlc != nil && tlc.BlockUntil == structs.TaskLifecycleBlockUntilCompleted {
+		onSuccess = false
+	} else if tlc != nil && tlc.BlockUntil == structs.TaskLifecycleBlockUntilRunning {
+		onSuccess = true
+	} else if jobType == structs.JobTypeBatch {
 		onSuccess = false
```

Comment:
You can simplify this logic a bit like with some comment explaining the rationale:

```golang
onSuccess := (jobType == structs.JobTypeService)
if tlc != nil && tls.Hook == structs.TaskLifecycleHookPrestart {
	onSuccess = (tlc.BlockUntil != structs.TaskLifecycleBlockUntilCompleted)
}
```

The conditionals above handles adding new block_until a tad better (e.g. when we add `healthy` as an option.